add `panic` keyword / add contributing section to Readme

inoas opened the pull request:
Sadly I don't know anything at all about treesitter:

```
❯ tree-sitter test
  imports:
    ✓ Imports
    ✓ Unqualified imports
    ✓ Aliased imports
  pipes:
    ✓ Pipes
  functions:
    ✓ Function examples
    ✓ Public function examples
    ✓ Basic functions
    ✓ Cases
    ✓ Try patterns
    ✓ Let expressions
    ✓ Complex binary expressions
    ✓ Complex nesting of field and tuple access
    ✓ Unusual function invocations
    ✓ Various discard variables
    ✓ Weird lists
    ✓ Comment in string
  targets:
    ✓ Target groups
    ✓ Target group edge cases
  custom_types:
    ✓ Parser example custom types
    ✓ Other custom type examples
    ✓ Public custom type definitions
    ✓ Public opaque custom type definitions
  whole_files:
    ✓ Excerpt from stdlib's base.gleam
    ✓ Excerpt from stdlib's bool.gleam
    ✓ Trailing commas
  external_functions:
    ✓ External functions
    ✓ Public external functions
  destructuring:
    ✓ Case with spread
  external_types:
    ✓ External types
    ✓ Public external types
  constants:
    ✓ Constants
    ✓ Public constants
    ✓ Scientific notation
  type_aliases:
    ✓ Type aliases
    ✓ Public type aliases
    ✓ Public opaque type aliases
  expressions:
    ✓ Bit-string expression
    ✓ Negation
    ✓ Concatenation
    ✓ Use
Error in query file "highlights.scm"

Caused by:
    Query error at 82:4. Invalid node type panic
```

Any hints?

J3RN commented on Cargo.toml:
```diff
@@ -1,7 +1,7 @@
 name = "tree-sitter-gleam"
 description = "gleam grammar for the tree-sitter parsing library"
-version = "0.26.2"
+version = "0.27.0"
```
Don't update the package version; I'll do that once all changes for v0.27.0 are in :grin:

J3RN commented on README.md:
```diff
@@ -23,22 +23,27 @@ trees and syntax highlighting for a given Gleam file.
-[tree-sitter-gleam-rust-example]: https://github.com/J3RN/tree-sitter-gleam-rust-example
-[tree-sitter-cli]: https://github.com/tree-sitter/tree-sitter/blob/master/cli/README.md
-[Install tree-sitter-cli]: https://github.com/tree-sitter/tree-sitter/blob/master/cli/README.md#installation
+[tree-sitter-gleam-rust-example]:
+https://github.com/J3RN/tree-sitter-gleam-rust-example
+[tree-sitter-cli]:
+https://github.com/tree-sitter/tree-sitter/blob/master/cli/README.md
+[Install tree-sitter-cli]:
+https://github.com/tree-sitter/tree-sitter/blob/master/cli/README.md#installation
```
I'm not a believer in wrapping Markdown content at a given line length.  The reason being that if you _don't_ wrap it, it can be read in a viewer/editor with line wrap at line length 60, 80, 100, etc; whatever the user wants.  If you _do_ wrap it and the user prefers a narrower line length (say, 60 characters), you get weird artifacting like this:

> 3. Clone this repository (or symlink it) into the new
`~/tree_sitters/`
directory.

J3RN commented on test/corpus/functions.txt:
```diff
@@ -359,6 +362,11 @@ fn field_access(x) {
+  (function
+    name: (identifier)
+    parameters: (function_parameters)
+    body: (function_body
+      (identifier)))
```
The body of this function is just `(panic)` :grin:
```suggestion
    body: (function_body
      (panic)))
```

J3RN commented on grammar.js:
```diff
@@ -341,6 +342,7 @@ module.exports = grammar({
       ),
     todo: ($) =>
       seq("todo", optional(seq("(", field("message", $.string), ")"))),
+		panic: (_$) => "panic",
```
GitHub is showing an indentation mismatch; when the project was started, I arbitrarily decided on two spaces for indentation, but this line has tab characters.

Running `npm run format` will automatically fix that (and other such things) :grin:

J3RN reviewed (commented):
Hey @inoas! :wave:  Thanks for submitting this patch! :tada:

To answer your direct question: the reason why the tests were failing is that the grammar needed to be "generated" from the `grammar.js` file by running `npm run generate`.  This is totally non-obvious and (AFAIK) not documented anywhere :sweat:   I'm making a change now such that when you run `npm test` it'll run `npm run generate` first to alleviate this issue!

I also flagged a handful of other tiny issues that we'll want to resolve before merging.

inoas commented:
The error is the same though:

```
Error in query file "highlights.scm"

Caused by:
    Query error at 82:4. Invalid node type panic
```

inoas commented:
> The error is the same though:
> 
> ```
> Error in query file "highlights.scm"
> 
> Caused by:
>     Query error at 82:4. Invalid node type panic
> ```

Seems like I've fixed it.

inoas commented:
I've tried to add `let assert` but the tests fail on that. The panic stuff seems to be working.

inoas commented:
So I have reverted the changes for `let assert` and just did the ones for panic.

inoas commented:
@J3RN could you triggrer the workflow?

inoas commented:
`npm run test` is green for me

inoas commented:
> Hey @inoas! 👋 Thanks for submitting this patch! 🎉
> 
> To answer your direct question: the reason why the tests were failing is that the grammar needed to be "generated" from the `grammar.js` file by running `npm run generate`. This is totally non-obvious and (AFAIK) not documented anywhere 😓 I'm making a change now such that when you run `npm test` it'll run `npm run generate` first to alleviate this issue!
> 
> I also flagged a handful of other tiny issues that we'll want to resolve before merging.


How do I emperically test?

> ❯ tree-sitter parse test/gleam/list_test.gleam
> No language found
> ❯ tree-sitter highlight test/gleam/list_test.gleam
> No language found for path "test/gleam/list_test.gleam"

I have got tree-sitter cloned

```
❯ which tree-sitter
/usr/local/bin/tree-sitter
```

I have this repo here cloned: `~/tree-sitters/tree-sitter-gleam`

the-mikedavis commented on README.md:
```diff
@@ -67,6 +68,14 @@ To-do List
+Contributing
+------------
+
+1. Change files such as `grammar.json` and `queries/highlight.scm`.
```
```suggestion
1. Change files such as `grammar.js` and `queries/highlight.scm`.
```

the-mikedavis commented:
You can use `npx tree-sitter parse path/to/file.gleam` to use the tree-sitter-cli version from the `package.json`. `./scripts/integration_test.sh` will run the integration tests against the gleam stdlib and other gleam repositories. It looks like the integration tests fail until #48 is merged in since `let assert` is used in the stdlib.

J3RN commented on README.md:
```diff
@@ -67,6 +68,14 @@ To-do List
+
+1. Change files such as `grammar.json` and `queries/highlight.scm`.
+2. The grammar needs to be generated from the `grammar.js` file by running `npm run generate`.
+3. Add lang feature tests in `test/corpus/functions.txt`.
+4. Run `npm run test` and fix tests in `test/corpus/functions.txt`.
```
`functions.txt` isn't the _only_ place for tests; there's a bunch of files in `test/corpus/` that exist for testing different bits of the language.  How about this?

```suggestion
3. Add parser feature tests to the relevant file(s) in `test/corpus/` (or make a new one!).
4. Run `npm run test` and fix any failing tests.
```

J3RN commented:
When updating this branch, don't worry about resolving the conflicts in `src/parser.c` by hand.  You can "resolve" the conflicts however you like&mdash;even just leaving the `>>>>` etc in there&mdash;and generate correct contents `src/parser.c` by running `npm run generate`.

J3RN reviewed (approved):
Looks great to me! :tada:  Thank you for your contribution! :purple_heart:

inoas commented:
> ./scripts/integration_test.sh

I do not understand this:

```
❯ git checkout main
Switched to branch 'main'
Your branch is up to date with 'origin/main'.
❯ ./scripts/integration_test.sh
Running parser against gleam-lang/stdlib
test/integration/gleam_lang_stdlib/src/gleam/string.gleam              	1 ms	(ERROR [859, 4] - [1025, 28])
test/integration/gleam_lang_stdlib/test/gleam/bool_test.gleam          	1 ms	(ERROR [159, 13] - [159, 14])
test/integration/gleam_lang_stdlib/test/gleam/iterator_test.gleam      	1 ms	(ERROR [0, 0] - [574, 0])
test/integration/gleam_lang_stdlib/test/gleam/list_test.gleam          	14 ms	(ERROR [848, 35] - [848, 41])
test/integration/gleam_lang_stdlib/test/gleam/queue_test.gleam         	3 ms	(ERROR [85, 2] - [172, 22])
test/integration/gleam_lang_stdlib/test/gleam/regex_test.gleam         	0 ms	(ERROR [5, 2] - [85, 67])
test/integration/gleam_lang_stdlib/test/gleam/string_test.gleam        	3 ms	(ERROR [415, 2] - [983, 50])
test/integration/gleam_lang_stdlib/test/gleam/uri_test.gleam           	1 ms	(ERROR [9, 2] - [554, 66])
Total parses: 45; successful parses: 37; failed parses: 8; success percentage: 82.22%

❯ git checkout add-panic
Switched to branch 'add-panic'
❯ ./scripts/integration_test.sh
Running parser against gleam-lang/stdlib
test/integration/gleam_lang_stdlib/src/gleam/string.gleam              	2 ms	(ERROR [859, 4] - [1025, 28])
test/integration/gleam_lang_stdlib/test/gleam/bool_test.gleam          	1 ms	(ERROR [159, 13] - [159, 14])
test/integration/gleam_lang_stdlib/test/gleam/iterator_test.gleam      	1 ms	(ERROR [0, 0] - [574, 0])
test/integration/gleam_lang_stdlib/test/gleam/list_test.gleam          	13 ms	(ERROR [848, 35] - [848, 41])
test/integration/gleam_lang_stdlib/test/gleam/queue_test.gleam         	4 ms	(ERROR [85, 2] - [172, 22])
test/integration/gleam_lang_stdlib/test/gleam/regex_test.gleam         	0 ms	(ERROR [5, 2] - [85, 67])
test/integration/gleam_lang_stdlib/test/gleam/string_test.gleam        	3 ms	(ERROR [415, 2] - [983, 50])
test/integration/gleam_lang_stdlib/test/gleam/uri_test.gleam           	1 ms	(ERROR [9, 2] - [554, 66])
Total parses: 45; successful parses: 37; failed parses: 8; success percentage: 82.22%
```

Why is it only at 82.22%?
At first I thought my panic change reduced that number

inoas commented:
ah ok, probably https://github.com/gleam-lang/tree-sitter-gleam/pull/48 was the source of the issue. With that rebased on main it is back at 100%.

Cool!
